qubes-decoration-handle-guivm-windows-prefix.patch

```diff
@@ -0,0 +1,55 @@
+From 96196938924004ebe7fa6890e17fb14c3d608e4b Mon Sep 17 00:00:00 2001
+From: =?UTF-8?q?Fr=C3=A9d=C3=A9ric=20Pierret=20=28fepitre=29?=
+ <frederic.pierret@qubes-os.org>
+Date: Tue, 5 Nov 2019 15:43:06 +0100
+Subject: [PATCH] Qubes Decoration: handle GuiVM windows prefix
+
+---
+ src/client.c | 23 ++++++++++++++++++++---
+ 1 file changed, 20 insertions(+), 3 deletions(-)
+
+diff --git a/src/client.c b/src/client.c
+index f2f366c6..ba965910 100644
+--- a/src/client.c
++++ b/src/client.c
+@@ -248,6 +248,23 @@ getQubesVmName (Client *c)
+         unsigned long nitems = 0, bytes_left;
+         char *data = 0;
+ 
++        // In case of GuiVM get qubes_vmname from root window if set
++        if ((XGetWindowProperty (display_info->dpy, XDefaultRootWindow(display_info->dpy), atom_vmname, 0L, 0L,
++                        FALSE, AnyPropertyType, &actual_type, &actual_format, &nitems,
++                        &bytes_left, (unsigned char **) &data) == Success))
++        {
++            if (bytes_left > 0) {
++                if ((XGetWindowProperty (display_info->dpy, XDefaultRootWindow(display_info->dpy), atom_vmname, 0L, bytes_left,
++                                FALSE, AnyPropertyType, &actual_type, &actual_format, &nitems,
++                                &bytes_left, (unsigned char **) &data) == Success))
++                {
++                    g_free (c->qubes_vmname);
++                    c->qubes_vmname = g_strdup_printf (_("[%s] "), data);
++                }
++            }
++        }
++
++        // Default case for getting qubes_vmname
+         if ((XGetWindowProperty (display_info->dpy, c->window, atom_vmname, 0L, 0L,
+                         FALSE, AnyPropertyType, &actual_type, &actual_format, &nitems,
+                         &bytes_left, (unsigned char **) &data) == Success))
+@@ -261,10 +278,10 @@ getQubesVmName (Client *c)
+                 } else {
+                     c->qubes_vmname = g_strdup (_("[ERROR Reading VM name?!] "));
```

Comment:
This will will still leak the memory - `c->qubes_vmname` assigned from the root window isn't freed here.

Generally better structure it differently:
Try to get vmname from the window itself (`c->window`). If that failed (`c->qubes_vmname` is still `NULL`), try from root window. If that still failed, then fallback to static `[Dom0]`.
This will avoid retrieving property from the root window if it isn't going to be used.